pkg/ruler/ruler.go

```diff
@@ -153,52 +154,58 @@ func NewRuler(cfg Config, engine *promql.Engine, queryable promStorage.Queryable
 		pusher:       pusher,
 		mapper:       newMapper(cfg.RulePath, logger),
 		userManagers: map[string]*promRules.Manager{},
-		done:         make(chan struct{}),
-		terminated:   make(chan struct{}),
 		registry:     reg,
 		logger:       logger,
 	}
 
+	ruler.Service = services.NewBasicService(ruler.starting, ruler.run, ruler.stopping)
+	return ruler, nil
+}
+
+func (r *Ruler) starting(ctx context.Context) error {
 	// If sharding is enabled, create/join a ring to distribute tokens to
 	// the ruler
-	if cfg.EnableSharding {
-		lifecyclerCfg := cfg.Ring.ToLifecyclerConfig()
-		ruler.lifecycler, err = ring.NewLifecycler(lifecyclerCfg, ruler, "ruler", ring.RulerRingKey, true)
+	if r.cfg.EnableSharding {
+		lifecyclerCfg := r.cfg.Ring.ToLifecyclerConfig()
+		var err error
+		r.lifecycler, err = ring.NewLifecycler(lifecyclerCfg, r, "ruler", ring.RulerRingKey, true)
 		if err != nil {
-			return nil, err
+			return errors.Wrap(err, "failed to initialize ruler's lifecycler")
 		}
 
-		ruler.lifecycler.Start()
+		r.ring, err = ring.New(lifecyclerCfg.RingConfig, "ruler", ring.RulerRingKey)
+		if err != nil {
+			return errors.Wrap(err, "failed to initialize ruler's ring")
+		}
 
-		ruler.ring, err = ring.New(lifecyclerCfg.RingConfig, "ruler", ring.RulerRingKey)
+		r.servManager, err = services.NewManager(r.lifecycler, r.ring)
 		if err != nil {
-			return nil, err
+			return err
 		}
+		err = r.servManager.StartAsync(ctx)
+		if err == nil {
+			err = r.servManager.AwaitHealthy(ctx)
+		}
+		return errors.Wrap(err, "failed to start ruler's services")
 	}
 
-	go ruler.run()
-	level.Info(logger).Log("msg", "ruler up and running")
-
-	return ruler, nil
+	// TODO: ideally, ruler would wait until its queryable is finished starting.
```

Comment:
This is another goal of this work, but outside of this PR... ruler should not start until queryable is ready (and in turn its store queryable, which in case of block store is fetching metadata from blocks). I think it doesn't make sense for ruler to start only to get bunch of errors.